Lib9c/Action/BattleGrandFinale.cs

```diff
@@ -0,0 +1,230 @@
+                    typeof(EquipmentItemOptionSheet),
+                    typeof(MaterialItemSheet),
+                });
+
+            avatarState.ValidEquipmentAndCostume(costumes, equipments,
+                sheets.GetSheet<ItemRequirementSheet>(),
+                sheets.GetSheet<EquipmentItemRecipeSheet>(),
+                sheets.GetSheet<EquipmentItemSubRecipeSheetV2>(),
+                sheets.GetSheet<EquipmentItemOptionSheet>(),
+                context.BlockIndex, addressesHex);
+
+            var grandFinaleSheet = sheets.GetSheet<GrandFinaleScheduleSheet>();
+            if (!grandFinaleSheet.TryGetValue(grandFinaleId, out var grandFinaleRow))
+            {
+                throw new SheetRowNotFoundException(nameof(GrandFinaleScheduleSheet),
+                    $"grandFinale Id : {grandFinaleId}");
+            }
+
+            if (!grandFinaleRow.IsOpened(context.BlockIndex))
+            {
+                throw new ThisArenaIsClosedException(
+                    $"{nameof(BattleGrandFinale)} : block index({context.BlockIndex}) - ");
+            }
+
+            var grandFinaleParticipantsSheet = sheets.GetSheet<GrandFinaleParticipantsSheet>();
+            if (!grandFinaleParticipantsSheet.TryGetValue(grandFinaleId, out var participantsRow))
+            {
+                throw new SheetRowNotFoundException(nameof(GrandFinaleParticipantsSheet),
+                    $"grandFinale Id : {grandFinaleId}");
+            }
+
+            if (!participantsRow.Participants.Contains(myAvatarAddress))
+            {
+                throw new AddressNotFoundInArenaParticipantsException(
+                    $"[{nameof(BattleGrandFinale)}] my avatar address : {myAvatarAddress}");
+            }
+
+            if (!participantsRow.Participants.Contains(enemyAvatarAddress))
+            {
+                throw new AddressNotFoundInArenaParticipantsException(
+                    $"[{nameof(BattleGrandFinale)}] enemy avatar address : {enemyAvatarAddress}");
+            }
+
+            var myArenaAvatarStateAdr = ArenaAvatarState.DeriveAddress(myAvatarAddress);
+            if (!states.TryGetArenaAvatarState(myArenaAvatarStateAdr, out var myArenaAvatarState))
+            {
+                throw new ArenaAvatarStateNotFoundException(
+                    $"[{nameof(BattleGrandFinale)}] my avatar address : {myAvatarAddress}");
+            }
+
+            var enemyArenaAvatarStateAdr = ArenaAvatarState.DeriveAddress(enemyAvatarAddress);
+            if (!states.TryGetArenaAvatarState(
+                    enemyArenaAvatarStateAdr,
+                    out var enemyArenaAvatarState))
+            {
+                throw new ArenaAvatarStateNotFoundException(
+                    $"[{nameof(BattleArena)}] enemy avatar address : {enemyAvatarAddress}");
+            }
+
+            var scoreAddress = myAvatarAddress.Derive(string.Format(CultureInfo.InvariantCulture, ScoreDeriveKey, grandFinaleId));
```

Comment:
Is there a case only the used score without the arena information?

Reply:
Come to think of it, it's right to include it in GrandFinaleInformation. I'll reflect it.